arrows/serialize/json/load_save.cxx

```diff
@@ -229,16 +229,29 @@ void load( ::cereal::JSONInputArchive& archive, kwiver::vital::detected_object_t
 void save( ::cereal::JSONOutputArchive& archive, const kwiver::vital::image_container_sptr ctr )
 {
   kwiver::vital::image vital_image = ctr->get_image();
+  kwiver::vital::image local_image;
+
+  if ( vital_image.memory() == nullptr || ! vital_image.is_contiguous() )
+  {
+    // Either we do not own the memory or it is not contiguous.  We
+    // need to consolidate the input image into a contiguous memory
+    // block before it can be serialized.
+    local_image.copy_from( vital_image );
+  }
+  else
+  {
+    local_image = vital_image;
+  }
 
   // Compress raw pixel data
   const uLongf size = compressBound( vital_image.size() );
   uLongf out_size(size);
   std::vector<uint8_t> image_data( size );
   Bytef* out_buf = reinterpret_cast< Bytef* >( &image_data[0] );
-  Bytef const* in_buf = reinterpret_cast< Bytef * >(vital_image.memory()->data());
+  Bytef const* in_buf = reinterpret_cast< Bytef * >( local_image.first_pixel() );
 
   int z_rc = compress( out_buf, &out_size, // outputs
-                       in_buf, vital_image.size() ); // inputs
+                       in_buf, local_image.size() ); // inputs
```

Comment:
There is still one case that is not handled correctly by using `size()` here.  `size()` is currently defined as `memory()->size()` if `memory() != nullptr` else `0`.  There is a case where the image memory is still contiguous but does not use the full memory (e.g. I adjust `first_pixel` and `height` to crop off a few rows of pixels from the start).  In this case you will be copying more data than needed and it will extend beyond the allocated memory.  The correct solution is to use `width * height * depth * pixel_format.num_bytes` instead of `size()`.

There is some argument that says this is what `size()` should be returning.  But that would give the wrong impression if someone copied `size()` bytes for a non-contiguous image.  Another option is that `size()` could return the number of bytes spanning the first pixel to the last pixel, even if not all pixels in between are seen.  What do you think?